pm_kext/src/pm_packet.c

```diff
@@ -0,0 +1,653 @@
+/*
+ *  Name:        pm_packet_utils.c
+ *
+ *  Owner:       Safing ICS Technologies GmbH
+ *
+ *  Description: Helper packet functions
+ *
+ *  Scope:       Kernelmode
+ */
+
+#include "pm_packet.h"
+#include "pm_checksum.h"
+#include "pm_netbuffer.h"
+#include "pm_utils.h"
+#include "pm_debug.h"
+
+static size_t getTCPResetPacketSizeIPv4();
+static size_t getTCPResetPacketSizeIPv6();
+static size_t getICMPBlockedPacketSizeIPv4(void *originalPacket, size_t originalPacketLength);
+static size_t getICMPBlockedPacketSizeIPv6(size_t originalPacketLength);
+
+static void generateTCPResetPacketIPv4(void *originalPacket, size_t originalPacketLength, void *packet);
+static void generateTCPResetPacketIPv6(void *originalPacket, size_t originalPacketLength, void *packet);
+static void generateICMPBlockedPacketIPv4(void *originalPacket, size_t originalPacketLength, bool useLocalHost, void *icmpPacket);
+static void generateICMPBlockedPacketIPv6(void *originalPacket, size_t originalPacketLength, bool useLocalHost, void *icmpPacket);
+
+static NTSTATUS sendTCPResetPacket(PortmasterPacketInfo* packetInfo, void* originalPacket, size_t originalPacketLength);
+static NTSTATUS sendICMPBlockedPacket(PortmasterPacketInfo* packetInfo, void* originalPacket, size_t originalPacketLength, bool useLocalHost);
+
+static void freeAfterInject(void *context, NET_BUFFER_LIST *nbl, BOOLEAN dispatch_level);
+
+static HANDLE injectV4Handle = NULL;
+static HANDLE injectV6Handle = NULL;
+
+NTSTATUS initializeInjectHandles() {
+     // Create the packet injection handles.
+    NTSTATUS status = FwpsInjectionHandleCreate(AF_INET,
+            FWPS_INJECTION_TYPE_NETWORK,
+            &injectV4Handle);
+    if (!NT_SUCCESS(status)) {
+        ERR("failed to create WFP in4 injection handle", status);
+        return status;
+    }
+
+    status = FwpsInjectionHandleCreate(AF_INET6,
+            FWPS_INJECTION_TYPE_NETWORK,
+            &injectV6Handle);
+    if (!NT_SUCCESS(status)) {
+        ERR("failed to create WFP in6 injection handle", status);
+        return status;
+    }
+
+    return STATUS_SUCCESS;
+}
+
+void destroyInjectHandles() {
+    if (injectV4Handle != NULL) {
+        FwpsInjectionHandleDestroy(injectV4Handle);
+        injectV4Handle = NULL;
+    }
+
+    if (injectV6Handle != NULL) {
+        FwpsInjectionHandleDestroy(injectV6Handle);
+        injectV6Handle = NULL;
+    }
+}
+
+HANDLE getInjectionHandleForPacket(PortmasterPacketInfo *packetInfo) {
+    if (packetInfo->ipV6 == 0) {
+        return injectV4Handle;
+    } else{
+        return injectV6Handle;
+    }
+}
+
+NTSTATUS injectPacket(PortmasterPacketInfo *packetInfo, UINT8 direction, void *packet, size_t packetLength, bool forceSend) {
```

Comment:
`bool forceSend` parameter is new.

Bugfix: Used by ICMP inject we change the source and destination of the packet data but we leave the original ip addresses in the packetInfo struct (which is checked for loopback addresses)